tfhe/tests/test_fhe_number_ops.rs

```diff
@@ -0,0 +1,234 @@
+#![cfg(any(feature = "integer", feature = "shortint"))]
+
+//! For now, in this test file, we don't want to check results
+//! but rather check that short int, int, dyn short int, dyn int
+//! all overload the same operators.
+//!
+//! For each operator overloaded operator we want to support $ variants:
+//!
+//! lhs + rhs
+//! lhs + &rhs
+//! &lhs + rhs
+//! &lhs + &rhs
+use std::fmt::Debug;
+use std::ops::{Add, BitAnd, BitOr, BitXor, Div, Mul, Rem, Shl, Shr, Sub};
+use tfhe::prelude::{FheDecrypt, FheTryEncrypt};
+use tfhe::ClientKey;
+
+macro_rules! define_operation_test {
+    ($name:ident => ($trait:ident, $symbol:tt)) => {
+        fn $name<T>(lhs: T, rhs: T)
+        where
+            T: Clone + $trait<T, Output = T>,
+            T: for<'a> $trait<&'a T, Output = T>,
+            for<'a> &'a T: $trait<T, Output = T> + $trait<&'a T, Output = T>,
+        {
+            let _ = &lhs $symbol &rhs;
+
+            let _ = &lhs $symbol rhs.clone();
+
+            let _ = lhs.clone() $symbol &rhs;
+
+            let _ = lhs $symbol rhs;
+        }
+    };
+}
+
+/// We keep this to improve tests later
+#[allow(dead_code)]
+fn static_supports_all_add_ways<T, FheT>(lhs_clear: T, rhs_clear: T, client_key: &ClientKey)
+where
+    T: Add<T, Output = T> + Copy + Debug + PartialEq,
+    FheT: FheTryEncrypt<T, ClientKey> + FheDecrypt<T>,
+    FheT: Clone + Add<FheT, Output = FheT>,
+    FheT: for<'a> Add<&'a FheT, Output = FheT>,
+    for<'a> &'a FheT: Add<FheT, Output = FheT> + Add<&'a FheT, Output = FheT>,
+{
+    let lhs = FheT::try_encrypt(lhs_clear, client_key).unwrap();
+    let rhs = FheT::try_encrypt(lhs_clear, client_key).unwrap();
+
+    let expected = lhs_clear + rhs_clear;
+
+    let r = &lhs + &rhs;
+    let dec_r = r.decrypt(client_key);
+    assert_eq!(dec_r, expected);
+
+    let r = &lhs + rhs.clone();
+    let dec_r = r.decrypt(client_key);
+    assert_eq!(dec_r, expected);
+
+    let r = lhs.clone() + &rhs;
+    let dec_r = r.decrypt(client_key);
+    assert_eq!(dec_r, expected);
+
+    let r = lhs + rhs;
+    let dec_r = r.decrypt(client_key);
+    assert_eq!(dec_r, expected);
+}
+
+define_operation_test!(supports_all_add_ways => (Add, +));
+define_operation_test!(supports_all_sub_ways => (Sub, -));
+define_operation_test!(supports_all_mul_ways => (Mul, *));
+define_operation_test!(supports_all_div_ways => (Div, /));
+define_operation_test!(supports_all_bitand_ways => (BitAnd, &));
+define_operation_test!(supports_all_bitor_ways => (BitOr, |));
+define_operation_test!(supports_all_bitxor_ways => (BitXor, ^));
+
+fn supports_scalar_add_with_u8<T>(lhs: T, rhs: u8)
```

Comment:
all scalar operations discard results ?

Reply:
Yes as the top level comment explains, the goal of these test is to ensure we implement the operator consistently over the different variant possible (lhs + rh , lhs + &rhs,  &lhs + rhs,  &lhs + &rhs)